### PR TITLE
[Work in progress] Webscocket channel, Do not merge.

### PR DESCRIPTION
**Pull request by borissmidt**
## Changes
adds support to use a websocket like a managed channel/connection and reuse it for all requests. 
#1081

Todo:
- [x] Add websocketChannel to the go proxy
- [ ] Test the websocketChannel proxy, to verify that there is no resource leak.
- [ ] Backpressure between proxy and client (nice to have)
- [ ] Java proxy implementation (probably for another repo)
- [ ] Add integration tests.
- [x] Typescript: Add websocketChannel to the typescript client.
- [x] Typescript: Channel recovery. (how is this done in http2)
- [ ] Cleanup commits
- [x] max number of streams setting per websocket channel
- [ ] ???

The code currently has been tested at [this commit](https://github.com/improbable-eng/grpc-web/commit/9acf8244d00973529b8c9d54b3f6cc97f71c9753). More low level tests have to be added.

### CODE REVIEW

File: go/grpcwebproxy/main.go [hunk 1]
```diff
@@ -44,6 +44,8 @@ var (
 	websocketPingInterval = pflag.Duration("websocket_ping_interval", 0, "whether to use websocket keepalive pinging. Only used when using websockets. Configured interval must be >= 1s.")
 	websocketReadLimit    = pflag.Int64("websocket_read_limit", 0, "sets the maximum message read limit on the underlying websocket. The default message read limit is 32769 bytes.")
 
+	useWebsocketsChannel  = pflag.Bool("use_websockets_channels", false, "whether to use a chanllel over websocket transport layer (alpha)")
```
Comment (hexfusion) [inline go/grpcwebproxy/main.go:47]: minor: s/chanllel/channel

File: go/grpcweb/wrapper.go [hunk 2]
```diff
@@ -221,6 +236,40 @@ func (w *WrappedGrpcServer) IsGrpcWebRequest(req *http.Request) bool {
...
+	headers := make(http.Header)
+	for _, name := range w.allowedHeaders {
+		if values, exist := req.Header[name]; exist {
+			headers[name] = values
+		}
+	}
+
+	ctx, cancelFunc := context.WithCancel(req.Context())
+	defer cancelFunc()
+	NewWebsocketChannel(wsConn, w.handler, ctx)
```
Comment (hexfusion) [inline go/grpcweb/wrapper.go:264]: we need to show were this went :)

File: go/grpcweb/wrapper.go [hunk 3]
```diff
@@ -123,6 +123,15 @@ func (w *WrappedGrpcServer) ServeHTTP(resp http.ResponseWriter, req *http.Reques
 		resp.WriteHeader(http.StatusForbidden)
 		_, _ = resp.Write(make([]byte, 0))
 		return
+	} else if w.opts.enableWebsocketChannels && w.IsGrpcWebSocketChannelRequest(req) {
+		if w.websocketOriginFunc(req) {
+			if !w.opts.corsForRegisteredEndpointsOnly || w.isRequestForRegisteredEndpoint(req) {
+				w.HandleGrpcWebsocketRequest(resp, req)
```
Comment (hexfusion) [inline go/grpcweb/wrapper.go:129]: This should be `HandleGrpcWebsocketChannelRequest` right?

**Review by hexfusion — COMMENTED** [review]
Added a few notes on small nits I know you're this working n this. Going to implement this and test it tonight/tomorrow.

File: go/test/server_test.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+	serverResponse = &HelloReply{}
+	if err := proto.Unmarshal(response2.GetBody().GetData(), serverResponse); err != nil {
+		fmt.Errorf("error %v", err)
+	}
+	log.Infof("got server response2 %v", serverResponse)
+	c.Close(websocket.StatusNormalClosure, "")
+}
+
+func CompleteGrpc(ws *websocket.Conn, streamId uint32, ctx context.Context) error {
```
Comment (hexfusion) [inline go/test/server_test.go:149]: How is the stream id managed? I might of missed that.

Comment (borissmidt) [reply go/test/server_test.go:149]: The client will give streamid's back and has to manage it properly.
I would use an incremental id and assume that you don't do 4g requests. 
A max number of streams per channel might be useful as protection.

File: go/grpcweb/wrapper.go (hunk 3 above)
Comment (borissmidt) [reply go/grpcweb/wrapper.go:129]: Good catch.

**Comment by borissmidt** [comment]
It seems that yesterday i forgot to commit a lot of files, i now added the typescript code but i sitl have to add the integration test.

**Comment by borissmidt** [comment]
I really get stuck at the integration tests, i want to analyse what goes wrong, but there are no nice stack traces because of the webpack build.
I also have a lot of trouble to keep the build consistently working when i make some changes to the libary and then want to rerun the tests. (`npm link ../client/web-grpc` doesn't seem to work properly.)

I think the channel never opens and that is why the tests fail but i cannot debug it which makes the further development very hard.

**Comment by johanbrandhorst** [comment]
Hi Boris,

Sorry that you went through all this work, but I honestly don't think we can accept this large of a contribution to the code base at this stage. This code is mostly in maintenance mode and I'm not confident that we'd be able to maintain this new addition. It appears you're run into some trouble with our tests, and I'm sorry about that, but unfortunately they're one of the problems with the current code base.

**Comment by borissmidt** [comment]
Ah to bad, atleast i learned some typescript and go on these rainy days. 
I still believe that this would solve some of the grpc-web problems and increase the support on the browsers.
Maybe i better start a fork and name it 'grpc-websocket' because i'm 100% sure that grpc/grpc-web are not open to websocket implementations.

**Comment by borissmidt** [comment]
The last test run i had, it seems that it works:
[logging.log](https://github.com/improbable-eng/grpc-web/files/8112421/logging.log)

**Comment by johanbrandhorst** [comment]
You are of course welcome to fork the project and do your own development as long as you abide by the license: https://github.com/improbable-eng/grpc-web/blob/master/LICENSE.txt

**Comment by johanbrandhorst** [comment]
I'm going to close this PR, since we have no intention of landing it.

**Pull request closed without merging**